### ISSUE TITLE
Scheduling article Content

### ISSUE THREAD

**Issue by joewhitsitt** [uiowa/uiowa#723]
I think this is definitely blocked until we get away from poor man's cron, but Wordpress offers it out of the box and several clients have asked for it in the past.

This would need to tie into our moderation workflow if implemented.

**Update (bdickey)**
Created a new IS at https://github.com/uiowa/uiowa/issues/723#issuecomment-1758037258 and scoped to focus on scheduling publishing of articles for the service and scheduled unpublishing only for ITS for now. Scheduled upublishing for articles and adding scheduling functionality to Pages or other entity types can come later in separate issues. Scheduling blocks/sections is discussed in #5405

**Comment by joewhitsitt** [comment]
@richardbporter informed me that cron is currently running on a :00, :15, :30, :45 (15min interval), so as long as we restrict the scheduling input to 15 min we could implement something like this.

Our webform content block has a scheduling option but is not restricted on input.

**Comment by joewhitsitt** [comment]
This was brought up in terms of additional alerts functionality at the Transportation meeting today. https://www.drupal.org/project/scheduler_content_moderation_integration provides the needed moderation workflow solution if this was tied to content types like articles.

I see potential for "reusable bits" though #2096

**Comment by richardbporter** [comment]
Cron has been reduced to once an hour, offset by the application ever since #1782. Its run through a web request which is slower. A site could setup their own cron job and then add their site URL to the exclude option of the cron command. 

There is some cron docs at https://wiki.uiowa.edu/display/drupaldev/Acquia+Applications

**Comment by joewhitsitt** [comment]
So this probably isn't doable for all sites then?

**Comment by richardbporter** [comment]
I guess it depends on how frequent we need to run it and how long it currently takes. I could time it.

**Comment by richardbporter** [comment]
The uiowa application took about 9m30s to run. It has 155 sites. 

Another thing we might want to consider is breaking up cron. I'm not sure if thats still necessary in D8.

**Comment by richardbporter** [comment]
> A site could setup their own cron job and then add their site URL to the exclude option of the cron command.

This needs to be done without `drush cron` as that will cause deployment errors. I think a curl command to the site cron URL would work.

**Comment by joewhitsitt** [comment]
Does https://github.com/uiowa/uiowa/pull/3123 open this up again?

**Comment by richardbporter** [comment]
Yea hopefully down the line it'll be possible.

**Comment by richardbporter** [comment]
I timed uiowa again and it took 6m23s with 158 databases.

**Comment by joewhitsitt** [comment]
Progress! So maybe the idea of 15 or 30 min windows of scheduling content is a possibility that can scale.

**Comment by richardbporter** [comment]
Blocked by #3383

**Comment by pyrello** [comment]
Can we close this or #5405 as a duplicate? Also, it seems like this is not blocked anymore since we have drush cron in place.

**Comment by joewhitsitt** [comment]
I think scheduling content in this situation was things like publishing an article. I think #5405 is more specific to individual sections/blocks within layout builder pages

**Comment by pyrello** [comment]
> I think scheduling content in this situation was things like publishing an article. I think #5405 is more specific to individual sections/blocks within layout builder pages

Ah, yeah. I had not read the story closely enough.

**Comment by joewhitsitt** [comment]
> > A site could setup their own cron job and then add their site URL to the exclude option of the cron command.
> 
> This needs to be done without `drush cron` as that will cause deployment errors. I think a curl command to the site cron URL would work. https://github.com/uiowa/uiowa/issues/723#issuecomment-760499962

If I understand this correctly, our current method of offsets still allows a site specific schedule job that calls the curl instead of drush to trigger cron as long as we give the offset cron enough time to finish. This would also depend on which app the site is on because of where the app sits in the offset order.

>To run cron from outside the site, go to https://sitenow.uiowa.edu/cron/dY...
>Enable the Automated Cron module to allow cron execution at the end of a server response.

**Comment by briand44** [comment]
Grooming conversation... we should probe https://www.drupal.org/project/scheduler and https://www.drupal.org/project/scheduler_content_moderation_integration and understand if we want to move forward with those modules or something custom that is not based off cron.

**Comment by briand44** [comment]
Grooming conversation... Brian will create a new story for the alert feature scheduling and keep this in backlog for possible future scheduling for alerts or other default content types if the need is there. Sounds like we do this for the feature and not just Facilities because folks don't have to use this just because it is available. Alert feature does not currently have content moderation so we don't currently need https://www.drupal.org/project/scheduler_content_moderation_integration. We would also want to remove the field that is part of the Facilities split that is related to publishing date. Sized as medium.

**Comment by briand44** [comment]
ITS is interested in having the ability to schedule publishing/unpublishing of articles so that when creating articles that are time-dependent, they can schedule unpublishing of those so they don't display news that is no longer relevant.

**Comment by GaryRidgway** [comment]
If we want to reduce cron usage, we don't necessarily have to publish and unpublish articles. We could just have them always be published, but then have two fields with "Activate after MM/DD/YYYY", and "Expire after MM/DD/YYYY", then we just have some logic whenever lists are constructed to ensure the current date is between those two days. Then cron never runs, it is only ever worried about when a user queries the lists that are relevant.

We could then batch run cron if we want to actually unpublish articles at a certain time.

**Comment by briand44** [comment]
Grooming conversation... Concerns with the rolling hour delay of publishing/unpublishing when opening this up for articles service wide. Would need to think about communication.

**Comment by mjoneill** [comment]
Katie M from Nursing just reached out and asked if this feature was available (her internal contact was not reachable today). 

Sharing because if/when we do some future feature mining, she's someone who would be interested in this functionality.

**Comment by briand44** [comment]
@mjoneill she wants to be able to schedule publishing/unpublishing of articles or some other type of content?

**Comment by mjoneill** [comment]
> @mjoneill she wants to be able to schedule publishing/unpublishing of articles or some other type of content?

Good question. I only asked her this: "Would you be interested in this for News only, or for basic pages".  She said both, then went on to describe a scenario that sounded like what she really wanted was to unpublish/publish a container on a page on a given schedule.  For her, she was thinking about it as a new page...but the use case involved swapping out one banner component (I believe) for another on a given schedule.

Unrelated: She had not heard that we were working on a D10 upgrade and got super excited when she learned that it was going to come to them, etc...

**Comment by briand44** [comment]
^^ Sounds more similar to https://github.com/uiowa/uiowa/issues/5405

**Comment by briand44** [comment]
Grooming conversation... Enable publishing of articles for the service but not unpublishing. Enable publishing and unpublishing for ITS only for now until we see if there are additional requests for unpublishing. We can add to basic pages in a future slice. Need to make sure works with moderation https://github.com/uiowa/uiowa/issues/723#issuecomment-733148198.

Iowa Now embargo could conflict with this.

Should have help text about possible hour delay

Should include this in docs as part of slice.

**Comment by joewhitsitt** [comment]
Make sure there aren't more surprises with other patch article forms:

<img width="510" alt="Screenshot 2023-10-10 at 2 19 41 PM" src="https://github.com/uiowa/uiowa/assets/4663676/28c2d2da-b19f-4429-a876-83ef9923548f">

**Comment by briand44** [comment]
# Stories

As a SiteNow content editor, I would like the ability to schedule publishing of article content so that I can schedule articles to be published in the future

As an ITS site owner, I would like the ability to schedule un-publishing of article content so that time-dependent articles don't appear once they are no longer relevant.

# Additional Info

We discussed turning on scheduled publishing of articles for the service but only turning on the ability to schedule unpublishing for ITS until we get more requests for the ability to schedule unpublishing.

We'll want to make sure this works with moderation https://www.drupal.org/project/scheduler_content_moderation_integration

Should have help text about possible hour delay

Should include this in docs as part of slice.

Discussed Iowa Now embargo feature could conflict with this and that we should investigate other patched article types https://github.com/uiowa/uiowa/issues/723#issuecomment-1756085329

# Proposed Solution

- [x] Enable scheduled publishing for articles
- [x] Enable scheduled unpublishing for ITS articles
- [x] Make sure works with moderation https://www.drupal.org/project/scheduler_content_moderation_integration
- [x] Help text about possible hour delay
- [x] Document in our SiteNow docs
- [x] Seems like we should hide the seconds since we can't schedule with that granularity anyway admin/config/content/scheduler
- [x] Make sure there aren't more surprises with other patch article forms including Iowa Now embargo https://github.com/uiowa/uiowa/issues/723#issuecomment-1756085329
- [x] Clean up alerts split since dependency will move to default

**Comment by pyrello** [comment]
> As an ITS site owner, I would like the ability to schedule publishing and unpublishing of article content so that we can schedule articles to be published in the future and have time dependant articles no longer appear once they are no longer relevant.

I would amend this to:
- As a SiteNow content editor, I would like the ability to schedule publishing of article content so that I can schedule articles to be published in the future
- As an ITS site owner, I would like the ability to schedule un-publishing of article content so that time-dependent articles don't appear once they are no longer relevant.

**Comment by briand44** [comment]
>I would amend this...

Makes sense, done.

**Comment by joewhitsitt** [comment]
Added a draft to the documentation page: https://sitenow.uiowa.edu/node/126/latest

**Comment by joewhitsitt** [comment]
Updated doc draft with editor restriction info

**Comment by joewhitsitt** [comment]
Trying a bit more before we probe an alternative method https://github.com/uiowa/uiowa/issues/7106

**Comment by joewhitsitt** [comment]
Published the changes to https://sitenow.uiowa.edu/documentation/article-content-type
